# App development opportunities (iOS and Android)

## Goal
The goal of this project is to understand the feature that can make an app profitable in both iOS and Android Store, in order to make data-driven decision for a development of our next app.
We want to develop an app that it is in English, free to download and to install; the main source of revenue consists of in-app adds.


## Data and Methodology

For our analysis we use two dataset:
* [The first one](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) contains more than 7,000 Apple iOS mobile application details
* [The second one](https://www.kaggle.com/lava18/google-play-store-apps) contains approximately 10,0000 Android mobile application details

We are going to import this data set and have a first look at the sample. We then perfom data cleaning on both samples, in order to make sure to avoid redundant data and consider only in english and free apps.


## Exploring data

We start by opening and exploring our two dataset. To make them easier to explore, we create a function named `explore_data`. The  `explore_data` function:
* Takes in four parameters:
    * `dataset`, which is expected to be a list of lists.
    * `start` and `end`, which are both expected to be integers and represent the starting and the ending indices of a slice from the data set.
    * `rows_and_columns`, which is expected to be a Boolean and has `False` as a default argument. 
* Slice the data set using `dataset[start:end]`.
* Loops thorough the slice, and for eache iteration pronts a row and adds a new line adter that row using `print('\n')`.

Let's import the data set:

In [1]:
from csv import reader
# Open iOS Dataset
opened_file = open ('AppleStore.csv', encoding='utf8')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

#Open Android Dataset
opened_file = open ('googleplaystore.csv', encoding='utf8')
read_file = reader(opened_file)
android = list (read_file)
android_header = android [0]
android = android[1:]

Let's explore the dataset using the `explore_data` function on both datasets:

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
print (android_header)
explore_data(android, 0,3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


The Android data set has 10,841 observations and 13 columns. By looking at the columns, we consider the columns 'App', 'Category', 'Reviews', 'Install', 'Type, 'Price' and 'Genres' relevant for our analysis.
The Apple data set contains 7,197 observations and 13 columns. By looking at the columns, we consider the columns 'Track name', 'currency', 'price'. 'rating_count_tot' 'rating_count_ver' and 'prime_genre' relevant for our analysis. Please have a look at the [documentation](https://www.kaggle.com/lava18/google-play-store-apps) to have a better understanding of the data set and its columns

In [4]:
print (ios_header)
explore_data(ios, 0,3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


The Apple data set contains 7,197 observations and 13 columns. By looking at the columns, we consider the columns 'Track name', 'currency', 'price', 'rating_count_tot' 'rating_count_ver' and 'prime_genre' relevant for our analysis. Please have a look at the [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) to have a better understanding of the data set and its columns.

## Data cleaning

The Google Play [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) reports an error in the dataset: the entry 10472 has missing value for the colum 'Rating', generating a shift for the subsequent columns:

>10472 Life Made WI-Fi Touchscreen Photo Frame 1.9 19.0 3.0M 1,000+ Free 0 Everyone NaN February 11, 2018 1.0.19 4.0 and up NaN




In order to remove this error, we apply the `del` statement, which removes the a specified row:

In [5]:
#let's check the row 10472:
print(android [10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [6]:
#We apply the del statement as follows:
del (android[10472])

In [7]:
#we check the size of our dataset, to check if the row has been cancelled:
len (android)

10840

### Duplicates elimination from the Android dataset

If we read the discussion section of the Google data set, it is reported that some apps have duplicate entries. For instance, Instagram has four entries:

In [8]:
for i in android:
    name = i[0]
    if name == 'Instagram':
        print (i)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


We want to check and isolate all the apps that have duplicates inside the Google data set. In order to do that, we:
- Create two list: one for storing the name of duplicate apps, and one for storing the name of unique apps.
- Loop through the `Android` data set and for each iteration:
    - We save the app name to a variable named `name`
    - If `name` was already in the `unique_apps` list, we append `name` to the `duplicate_apps` list
    - Else, we appended `name` to the `unique_apps` list

Since we don't want to count certain apps more than once, when we analize data, we eliminate the duplicates entry according to the number of reviews of the app (contained in the "Reviews" column). We only keep the row with the highest number of reviews and remove the other entries for any given app.

In [9]:
#Android data set
duplicate_apps = []
unique_apps = []

for apps in android:
    name = apps[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print ('Number of duplicated apps: ', len(duplicate_apps))
print ('Number of unique apps: ', len(unique_apps))

Number of duplicated apps:  1181
Number of unique apps:  9659


In [10]:
#Apple data set
duplicate_apps = []
unique_apps = []

for apps in ios:
    name = apps[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print ('Number of duplicated apps: ', len(duplicate_apps))
print ('Number of unique apps: ', len(unique_apps))

Number of duplicated apps:  0
Number of unique apps:  7197


We see that in the Google Play data set, 1,181 observations are duplicates. After we remover this duplicates we should be left with 9,659 rows.
To remove these duplicates, we will:
- Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of the app
- Use the information stored in the dictionary and create a new dataset, which will have only one entry per app 

Let's start by creating the dictionary:

In [11]:
# Creation of the dictionary:
reviews_max = {}
for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name]<n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

We expect the lenght of the dicitionary equal to 9,659, since it only contains unique values



In [12]:
len(reviews_max)

9659

We use the dictionary we created to remove duplicates rows. In order to do that:
* We start by creating two lists: `android clean`, that stores the clean dataset and the `already_added` which will just store app names
* We loop through the Android data set and for each iteration:
    * Assign the app name to a variable named `name`
    * Convert the number of reviews to float, and assign it to a vairable named `n_reviews`.
* If `n_reviews` == number of reviews contained of the app `name` (contained in the dictionary), and name is not in `already_added`:
    * We append the entire row to the `android_clean` list
    * Append the name of the app `name` to the `already_added` list

In [13]:
android_clean = []
already_added = []
for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)

Here below we explore the cleaned dataset: as we expected, the numer of rows is 9.659

In [14]:
explore_data(android_clean, 1, 5, True)

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows: 9659
Number of columns: 13


### Removing non-Eglish apps

We use English for the apps we want to develop at our company and we would like to analyze only the apps that are directed toward an English-speaking audience. If we explore the data long enough, we will find that both data sets ahve apps with names that suggest thay are not directed toward an Enlgish-speaking audience.
We report below a couple of examples for each data set:

In [15]:
print (ios[813] [1])
print (ios [6731] [1])
print (android_clean [4412][0])
print (android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


We are not interested keeping these apps, so we are going to remove them.
One way to go about this is to remove each app with a name containing a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).

Behind the scenes, each character we use in a string has a corresponding number associated with it. For instance, the corresponding number for character `a` is 97, character `A` is 65, and character `爱` is 29,233. We can get the corresponding number of each character using `ord()`

- We first write a function that takes in a string and return `False` if there's any character in the string that doesn't belong to the set of common English characters, otherwise it returns `True`. 
- Inside the function, we iterate over the input string. For each iteration we check whether the number associated with the character is greater than 127. When a character is greater than 127, the function should immediately return `False` — the app name is probably non-English since it contains a character that doesn't belong to the set of common English characters.
- If the loop finishes running without the return statement being executed, then it means no character had a corresponding number over 127 — the app name is probably English, so the functions should return `True`

In [16]:
def is_english(string):
    for letter in string:
        if ord(letter) >127:
            return False
    return True

We check the correctness of our function:

In [17]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
False


However if we pass the string 'Instachat 😜' ,over function recognizes the app as non-english: this is because emojis fall outside the ASCII range and have corresponding values over 127

In [18]:
print(is_english('Instachat 😜'))

False


We need to modify our previous function, in order to avoid to lose useful data for our analysis. To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English. Our filter function is still not perfect, but it should be fairly effective.

In [19]:
def is_english(string):
    counter = 0
    for word in string:
        if ord(word) > 127:
            counter +=1
    if counter >3:
        return False
    return True

As we can see below, the function now works correctly:

In [20]:
print(is_english('Instachat 😜'))
print (is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
False


We use the `is_english` function to filter out non-english apps from both datasets. We loop through each dataset and if the app is identified as english we append the whole row to a separate list

In [21]:
#Filter for the Android and iOS data set:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name) == True:
        android_english.append(app)

for app in ios:
    name = app[1]
    if is_english(name) == True:
        ios_english.append(app)
        
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

### Selecting only free apps

As we mentioned at the beginning of the analysis, we need to isolate the free apps. In order to do that, we looop through each data set to separate the free apps in separate lists. 
The column for the Android data set that contain the Price has index 4, while for the iOS data set the index is 7:

In [22]:
android_final = []
ios_final = []

#iterate through the android dataset:
for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)

print(len(android_final))
print(len(ios_final)) 

8864
3222


The final dataset consists of 8,864 observations for the Android Dataset and 3,222 for the Apple Dataset 

## Most common Apps by genere

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.
Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets.

Let's begin the analysis by getting a sense of what are the most common genres for each market. For this, we'll need to build frequency tables for a few columns in our data sets.

The columns we are going to use to find out what are the most common genres in each markets are:
 -  'Prime_genre' for the iOS data set
 - 'Genres' and 'Category' for the Android data set

We build **two** functions to analyze the frequency tables:
1. One function to generate frequency tables that show percentage
2. Another function we can use to display the percentage in a descending order

We are able to build dictionaries to build frequencty tables, however dictionaries do not have orders, therefore we need to build the function n.2 described in the pargraph above.

- First of all we are going to build a `freq_table()` function, that comput the observed frequency of the apps sorted by genres.
- We then define a `display_table` function that:
    1. Takes in two parameters: `dataset` and `index`. `dataset` is expected to be a list of lists, and `index` is expected to be an integer.
    2. Generates a frequency table using the `freq_table()` function.
    3. Transforms the frequency table into a list of tuples, then sorts the list in a descending order.
    4. Prints the entries of the frequency table in descending order.

In [23]:

def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

We start our analysis by observing the iOS frequency table:

In [24]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


As we can see, the most common genre in the iOS data set is 'Gaming', representing 58% of the whole sample, followed by Entertainment apps (almost 8%) and photo and video apps (almost 5%). The 3 genres represent more thant 70% of the whole dataset, this however does not imply that these apps are also the largerst numbers of users/download.

We now analyze the Android data set:

In [25]:
display_table(android_final, 1) # We first look at the Category row

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The landscape is different on Google Play: there are not many apps designed for fun, a consistent number of them are designed for practical purposese (family, tools, business, lifestyle, productivity). However, by checking at the Google app store, the family category mostly include games for kids.

We now have the same analysis by considering the colum Genres

In [26]:
display_table(android_final, -4) # we now look aht the Genres column

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

Even if it is not clear the difference among Genre and Category rows, we are going to focus our analysis on Genre row, since the Category row is too granular for our analysis 

### Most popular Apps by Genre on the App Store



We are going to identfy the most popular apps by genre starting by computing the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

We calculate the average number of user ratings per app genre on the App Store

In [27]:
genres_ios = freq_table(ios_final, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


Navigation apps have on average the highest number of user reviews, but this number is influenced by Waze and Google Maps as we can see below. The same result applies also for the category Photo and Video, Games and Music

In [28]:
print ('----------Navigation------------')
for app in ios_final:
    if app[-5] == 'Navigation':
        print (app[1], ':', app[5])
print ('---------Photo & Video------------')

for app in ios_final:
    if app[-5] == 'Photo & Video':
        print (app[1], ':', app[5])
print ('---------Games------------')
for app in ios_final:
    if app[-5] == 'Games':
        print (app[1], ':', app[5])
print ('---------Music------------')        
for app in ios_final:
    if app[-5] == 'Music':
        print (app[1], ':', app[5])        

----------Navigation------------
Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5
---------Photo & Video------------
Instagram : 2161558
Snapchat : 323905
YouTube - Watch Videos, Music, and Live Streams : 278166
Pic Collage - Picture Editor & Photo Collage Maker : 123433
Funimate video editor: add cool effects to videos : 123268
musical.ly - your video social network : 105429
Photo Collage Maker & Photo Editor - Live Collage : 93781
Vine Camera : 90355
Google Photos - unlimited photo and video storage : 88742
Flipagram : 79905
Mixgram - Picture Collage Maker - Pic Photo Editor : 54282
Shutterfly: Prints, Photo Books, Cards Made Easy : 51427
Pic Jointer – Photo Collage, Camera Effects Editor : 51330
Color Pop Effects - Photo Editor & Picture Editing : 45320
Photo Grid - photo collage mak

Daily Celebrity Crossword : 181384
AdVenture Capitalist : 181359
Blackjack by MobilityWare : 180087
Call of Duty®: Heroes : 179416
Deer Hunter Classic : 179111
Agar.io : 176514
Piano Tiles (Don't Tap The White Tile) : 172686
Angry Birds Rio : 170843
The Secret Society® - Hidden Mystery : 168674
Real Racing 3 : 164483
Highway Rider : 164377
Words with Friends – Best Word Game : 160668
Bowmasters - Top Multiplayer Bowman Archery Game : 159323
aa : 158845
2048 : 157882
Episode - Choose Your Story : 155903
Dumb Ways to Die : 151559
Jurassic World™: The Game : 149138
Color Switch : 145043
Iron Force : 141634
Racing Penguin Free - Top Flying and Diving Game : 141224
Two Dots : 139486
Jigsaw Puzzle : 138726
Summoners War : 137122
Sniper Shooter: Gun Shooting Games : 134080
Candy Crush Soda Saga : 133394
My Singing Monsters : 130519
Spider Solitaire Free by MobilityWare : 128739
Smash Hit : 126233
▻Solitaire : 124618
My Talking Tom : 123859
Clumsy Ninja : 123032
Tetris® Blitz : 122868
Star War

WordBrain 2 : 4577
Blocky Football : 4549
Trump Dump ⁢ : 4544
Sandstorm: Pirate Wars : 4518
Tankio - Battle Arena : 4513
Shuffle Cats : 4499
Narcos: Cartel Wars : 4490
Craft your games! - Createrria 2 : 4475
Twisty Arrow - Shoot the Circle Wheel : 4440
Weed Firm 2: Back To College : 4387
Beach Buggy Racing : 4386
Rival Stars Basketball : 4383
Warrior Balls : 4375
Flick Quarterback 17 : 4371
Jam League : 4361
Descendants : 4333
Cut the Rope: Time Travel Free : 4327
Card Wars Kingdom - Adventure Time : 4283
Diamond Quest! : 4278
Invasion: Modern Empire : 4276
YO-KAI WATCH Wibble Wobble : 4275
Peanuts: Snoopy's Town Tale : 4260
Risky Road : 4226
NCIS: Hidden Crimes : 4222
Slash Mobs : 4179
Naughty Kitties : 4162
Land Sliders : 4142
Fun Run Arena - Online Multiplayer Running Game : 4116
Frozen Frenzy Mania: Challenging Match 3 Games : 4104
Oz: Broken Kingdom™ : 4074
Guess The Song - New music quiz! : 4050
Big Win Football 2016 : 4035
Loop Drive 2 : 4009
Dashy Crashy : 3979
Fireboy and Wate

Mahjöng : 1173
iSlash Heroes : 1170
Slots: Get Rich Slot Machines Casino Slot Games : 1168
Earn to Die 2 Lite : 1147
One More Dash : 1142
EA SPORTS™ FIFA 17 Companion : 1137
Frozen Free Fall: Icy Shot : 1133
Rival Stars College Football : 1132
Blocky Racer - Endless Arcade Racing : 1122
RAMBOAT - Game of Shoot and Dash, Run and Jump ! : 1114
Double Juggle : 1105
MultiCraft -- Survival Craft Build Sandbox Game : 1102
Slip Away : 1102
Dancing with the Stars: The Official Game : 1098
BlazBlue RR - the Real Action Game : 1091
Flappy Golf 2 : 1089
Stretch Dungeon : 1087
Smash Fu - Endless Arcade Smasher : 1086
Keepy Ducky : 1083
Brick Splits : 1077
Sky Chasers : 1074
Bike Traffic Rider an Extreme Real Endless Road Racer Racing Game : 1052
Harvest Land : 1047
Buddy Toss : 1018
Candy Party: Coin Carnival Dozer : 1017
Really Bad Chess : 1015
Pocket Troops : 1005
Lost in Harmony : 1002
PAC-MAN Pop : 1000
Tap Tap Builder : 999
Emoji Blitz : 999
Monster Tail : 998
Mr. Crab 2 : 989
ROCKY™ : 989
Tr

Fast Food Rampage : 281
Wire Bounce : 279
Muhammad Ali: Puzzle King : 277
Snail Bob 2 : 276
Paper Girl - Morning Madness Adventures : 274
CubeMator - Craft, Build and Explore the World : 273
Daddy's Little Helper - Messy Home Adventure : 270
NeoWars : 267
VR Roller Coaster for Google Cardboard : 266
Hot Pot Master : 265
Dragon Heroes: Shooter RPG : 265
Star Cheerleader - High School Tryouts : 263
Fruit Pop! Puzzles in Paradise - Fruit Pop Sequel : 261
Tooth Fairy Princess - Magical Adventure : 260
Learn 2 Fly : 258
Solitaire ▪ : 258
Maximum Car : 256
EatMe.io: Hungry Fish Attack! : 256
Magic Mansion : 255
WIND runner adventure : 253
Car Factory Parking Simulator a Real Garage Repair Shop Racing Game : 253
Gang City : 251
Stop Curry : 250
Supermodel Star - Rule the Runway : 250
SBK16 - Official Mobile Game : 249
High Dive : 249
Osmo Masterpiece : 247
Chaos Centurions : 244
Dropple : 243
Sloomy : 241
VR HORROR : 241
LEGO® Ninjago: Skybound : 239
Sniper: Traffic Hunter : 239
Tap the Block

My Rockstar Girls - Party Rock Band : 63
Splash Pop : 63
Exodus : 62
Spiral Splatter : 62
Push the Arrow : 62
Puzzlepops! Trick or Treat : 61
Nope Quiz : 61
White Tiles 2 : Piano Master ( Don't Hit The White Tiles 4 ) - Free Music Game : 60
Haunted Rooms: Escape VR Game for Google Cardboard : 60
Monkey Swingers : 60
Shoutrageous! - The Addictive Game of Lists : 60
Flying Car Simulator 3D: Stunt Bus : 59
Police Chase Race : 59
Gear Miner : 58
Cubway : 58
Indoor soccer – football Dream league journey : 58
Osmo Coding Awbie : 58
Gravity Square! : 58
Goal Kick - free penalty shootout soccer game : 57
Flick Rugby 16 : 57
Chef Siblings - Island Restaurant : 56
Triennale Game Collection : 56
Micromon : 56
Brio - Don’t Fall! : 56
Escape Game: Forgotten : 55
Dominocity : 55
Donuts Maker Salon : 55
Bouncy Kingdom : 55
ZenDots 2 - One Dot’s Journey : 54
Sweet Princess Beauty Salon : 54
Oops You Died : 54
Diced - A Simple Puzzle Dice Game : 53
Perfect Fit : 53
Color 6: Blitz : 53
Piggy Show : 53
S

My Mixtapez Music : 26286
Sing Karaoke Songs Unlimited with StarMaker : 26227
Ringtones for iPhone & Ringtone Maker : 25403
Musi - Unlimited Music For YouTube : 25193
AutoRap by Smule : 18202
Spinrilla - Mixtapes For Free : 15053
Napster - Top Music & Radio : 14268
edjing Mix:DJ turntable to remix and scratch music : 13580
Free Music - MP3 Streamer & Playlist Manager Pro : 13443
Free Piano app by Yokee : 13016
Google Play Music : 10118
Certified Mixtapes - Hip Hop Albums & Mixtapes : 9975
TIDAL : 7398
YouTube Music : 7109
Nicki Minaj: The Empire : 5196
Sounds app - Music And Friends : 5126
SongFlip - Free Music Streamer : 5004
Simple Radio - Live AM & FM Radio Stations : 4787
Deezer - Listen to your Favorite Music & Playlists : 4677
Ringtones for iPhone with Ringtone Maker : 4013
Bose SoundTouch : 3687
Amazon Alexa : 3018
DatPiff : 2815
Trebel Music - Unlimited Music Downloader : 2570
Free Music Play - Mp3 Streamer & Player : 2496
Acapella from PicPlayPost : 2487
Coach Guitar - Lessons

Reference apps have 74,942 users ratings on average, but actually the Bible and Dictionary skew the average rating

In [29]:
for app in ios_final:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


This niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

- Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

- Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

- Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

### Most popular Apps by Genre on Google Play


The data rearinf the genre popularity are stored in the install columns, however we can see that most of the values are open ended (1,000,000+, 1,000+, 50+..)

In [30]:
display_table(android_final, 5) # the Installs columns


1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).



In [31]:
categories_android = freq_table(android_final, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most install, skewed by few apps with over one billion installs and a few other with over 100 and 500 million installs:

In [32]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:



In [33]:
under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3603485.3884615386

The same pattern applies for PHOTOGRAPHY, SOCIAL, PRODUCTIVITY

In [34]:
for app in android_final:
    if app[1] == 'PHOTOGRAPHY' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

B612 - Beauty & Filter Camera : 100,000,000+
YouCam Makeup - Magic Selfie Makeovers : 100,000,000+
Sweet Selfie - selfie camera, beauty cam, photo edit : 100,000,000+
Google Photos : 1,000,000,000+
Retrica : 100,000,000+
Photo Editor Pro : 100,000,000+
BeautyPlus - Easy Photo Editor & Selfie Camera : 100,000,000+
PicsArt Photo Studio: Collage Maker & Pic Editor : 100,000,000+
Photo Collage Editor : 100,000,000+
Z Camera - Photo Editor, Beauty Selfie, Collage : 100,000,000+
PhotoGrid: Video & Pic Collage Maker, Photo Editor : 100,000,000+
Candy Camera - selfie, beauty camera, photo editor : 100,000,000+
YouCam Perfect - Selfie Photo Editor : 100,000,000+
Camera360: Selfie Photo Editor with Funny Sticker : 100,000,000+
S Photo Editor - Collage Maker , Photo Collage : 100,000,000+
AR effect : 100,000,000+
Cymera Camera- Photo Editor, Filter,Collage,Layout : 100,000,000+
LINE Camera - Photo editor : 100,000,000+
Photo Editor Collage Maker Pro : 100,000,000+


In [35]:
for app in android_final:
    if app[1] == 'SOCIAL' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Facebook : 1,000,000,000+
Facebook Lite : 500,000,000+
Tumblr : 100,000,000+
Pinterest : 100,000,000+
Google+ : 1,000,000,000+
Badoo - Free Chat & Dating App : 100,000,000+
Tango - Live Video Broadcast : 100,000,000+
Instagram : 1,000,000,000+
Snapchat : 500,000,000+
LinkedIn : 100,000,000+
Tik Tok - including musical.ly : 100,000,000+
BIGO LIVE - Live Stream : 100,000,000+
VK : 100,000,000+


In [36]:
for app in android_final:
    if app[1] == 'PRODUCTIVITY' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Microsoft Word : 500,000,000+
Microsoft Outlook : 100,000,000+
Microsoft OneDrive : 100,000,000+
Microsoft OneNote : 100,000,000+
Google Keep : 100,000,000+
ES File Explorer File Manager : 100,000,000+
Dropbox : 500,000,000+
Google Docs : 100,000,000+
Microsoft PowerPoint : 100,000,000+
Samsung Notes : 100,000,000+
SwiftKey Keyboard : 100,000,000+
Google Drive : 1,000,000,000+
Adobe Acrobat Reader : 100,000,000+
Google Sheets : 100,000,000+
Microsoft Excel : 100,000,000+
WPS Office - Word, Docs, PDF, Note, Slide & Sheet : 100,000,000+
Google Slides : 100,000,000+
ColorNote Notepad Notes : 100,000,000+
Evernote – Organizer, Planner for Notes & Memos : 100,000,000+
Google Calendar : 500,000,000+
Cloud Print : 500,000,000+
CamScanner - Phone PDF Creator : 100,000,000+


We now look at the BOOKS and REFERENCE category to see if this genre could be also interesting for the Android Store

In [37]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

Even in this case, we find out that a small number of apps are extremely popular:

In [38]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


By looking at only the mid-size popular app, we can see that few apps are built around the book Quran, which suggests that building an app around a popular book can be profitable:

In [39]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

# Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc